### PR TITLE
[Merged by Bors] - Allow vaults to spend received funds

### PR DESCRIPTION
**Pull request by lrettig**
## Motivation

Fixes an oversight in genvm implementation: vesting vaults are unable to spend received coins.

## Description

Adds a `Balance()` method to host/context to allow genvm templates to read current account balance, then updates vault spending arithmetic accordingly.

## Test Plan

Added tests in `genvm/templates/vault/vault_test.go`

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed
- [x] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

**Review by fasmat — APPROVED** [review]

**Comment by lrettig** [comment]
thanks @fasmat. let's get a second review on this one since the code path is pretty critical.

**Comment by lrettig** [comment]
@dshulyak pointed out that there's nothing currently preventing a misconfigured vault (with, say, an actual balance less than the `TotalAmount`) from being spawned. I addressed this, please see the latest commits.

This however raises the question - do we want to allow Spawning of misconfigured/invalid Vaults in the first place? Examples would be an insufficient balance, or `VestingStart > VestingEnd`. With the code as it stands there's no harm in allowing them, and I've written the logic here in such a way that the funds in the vault can still be spent. We may also want to prevent Spawn of any Vault other than the genesis Vaults to prevent this sort of issue.

If `VestingStart > VestingEnd`, no Spend can occur until `VestingStart` layer, at which point the `TotalAmount` becomes available.
If account balance < unvested portion of `TotalAmount`, no Spend can occur, but this can always be remedied by sending more coins to the account.

File: genvm/templates/vault/vault.go
Comment (poszu) [inline genvm/templates/vault/vault.go]: TBH, I don't understand how it works now. Why deduct the total amount (max vault worth when fully vested) here? Does `host.Balance()` contain the "future/unvested" `TotalAmount` in it?

File: genvm/templates/vault/vault.go
Comment (poszu) [inline genvm/templates/vault/vault.go:62]: `DrainedSoFar` is not used anymore - could you remove it?

Comment (lrettig) [reply genvm/templates/vault/vault.go:62]: good catch. since the balance check now uses the actual account balance, this is no longer necessary.

File: genvm/templates/vault/vault.go
Comment (lrettig) [reply genvm/templates/vault/vault.go]: > Does host.Balance() contain the "future/unvested" TotalAmount in it?

yup

I made the comment a little more explanatory:

	// current account balance minus unvested portion of initial endowment equals unspent, vested coins
	// plus coins received

Is that clearer? Here's pseudocode:

```
unvested_portion = v.TotalAmount - vested
spendable_portion = host.balance() - unvested_portion
if amount > spendable_portion { ... }
```

Comment (poszu) [reply genvm/templates/vault/vault.go]: Yes, thank you! 

A NIT: Consider putting that pseudocode in the code instead of the comment. IMHO, it's better if the code can explain itself. ;)

**Comment by poszu** [comment]
> This however raises the question - do we want to allow Spawning of misconfigured/invalid Vaults in the first place? Examples would be an insufficient balance, or `VestingStart > VestingEnd`. With the code as it stands there's no harm in allowing them, and I've written the logic here in such a way that the funds in the vault can still be spent. We may also want to prevent Spawn of any Vault other than the genesis Vaults to prevent this sort of issue.

I prefer when objects in invalid state cannot exist in the first place rather than have to have checks around in the logic because the checks are easy to miss or get wrong.

> If `VestingStart > VestingEnd`, no Spend can occur until `VestingStart` layer, at which point the `TotalAmount` becomes available. If account balance < unvested portion of `TotalAmount`, no Spend can occur, but this can always be remedied by sending more coins to the account.

I think `VestingStart > VestingEnd` is not possible because of this check: https://github.com/spacemeshos/go-spacemesh/blob/e2c579ad9322a148b99ca62c1bfdb2aa5b392fc5/genvm/templates/vault/handler.go#L42-L45

However, `VestingStart == VestingEnd` is possible, in which case the code in https://github.com/spacemeshos/go-spacemesh/blob/e2c579ad9322a148b99ca62c1bfdb2aa5b392fc5/genvm/templates/vault/vault.go#L49 would crash (division by zero).

**Comment by lrettig** [comment]
> I prefer when objects in invalid state cannot exist in the first place rather than have to have checks around in the logic because the checks are easy to miss or get wrong.

Makes sense. But the question is orthogonal to the fixes in this PR. Let's get this merged and then address that question separately.

> However, `VestingStart == VestingEnd` is possible, in which case the code in
> 
> https://github.com/spacemeshos/go-spacemesh/blob/e2c579ad9322a148b99ca62c1bfdb2aa5b392fc5/genvm/templates/vault/vault.go#L49
> 
> would crash (division by zero).

Actually division by zero is impossible here. If `VestingStart == VestingEnd` then one of these two conditionals will be triggered first:

https://github.com/spacemeshos/go-spacemesh/blob/e2c579ad9322a148b99ca62c1bfdb2aa5b392fc5/genvm/templates/vault/vault.go#L41-L46

Note that there's already a test for this:

https://github.com/spacemeshos/go-spacemesh/blob/e2c579ad9322a148b99ca62c1bfdb2aa5b392fc5/genvm/templates/vault/vault_test.go#L304-L312

I'll add a comment in the code to this effect.

**Review by poszu — APPROVED** [review]

**Comment by lrettig** [comment]
bors merge

**Comment by lrettig** [comment]
bors merge

**Comment by lrettig** [comment]
bors merge

**Pull request closed without merging**
